### PR TITLE
Improve nightly functional tests

### PR DESCRIPTION
**Pull request by james-garriss**
## 🗣 Description ##

There are several improvements to the nightly functional tests that are all rolled into this PR:

* Publish to the private gallery, thus ensuring that our code is in a state that can be published to PSGallery.
* Email TCOs for each failing set of tests, thus allowing them to know when there are issues that may need their attention.
* Use GitHub caching, thus speeding up testing time by only installing GitHub and various dependencies once.
* Handle concurrency problems within a run by using GitHub's "needs" features for product/tenant pairing, thus ensuring testing one variant doesn't mess with testing a different variant.
* Handle concurrency problems between runs by using GitHub's "concurrency" features for product/tenant pairing, thus ensuring that one tester doesn't mess with the tests of another tester.
* Added every combination of products/tenants/variants, thus ensuring more complete coverage.  This required additional test parameters for both dev and prod.

The matrix solution that I was using doesn't support concurrency, so I had to replace it with a set of reusable workflows and composite actions.  The resulting solution violates DRY more than I want, so I may eventually create an exploratory issue to replace the reusable workflows with composite action.

I also fixed a permissions problem with Checkov and fixed a pathing problem with the CI and workflow pipelines, as they were impacting my work.

## 💭 Motivation and context ##

The nightly functional tests were good, but they lacked some features that were causing issues (e.g., concurrency), and they were incomplete.

Closes: #1235 
Closes: #1237 
Closes: #1278 

## 🧪 Testing ##

Development was initially done w a temp pipeline whose runs can be found here:
https://github.com/cisagov/ScubaGear/actions/workflows/test_concurrency.yaml
Note: At some point these runs will be cleaned up and deleted.

Once the work was completed, the code was transitioned to the nightly workflow:
https://github.com/cisagov/ScubaGear/actions/workflows/test_production_function.yaml

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] PR targets the correct parent branch (e.g., main or release-name) for merge.
- [x] Changes are limited to a single goal - *eschew scope creep!*

Well, I did clean up some other issues, too, but they were impacting my testing.

- [x] Changes are sized such that they do not touch excessive number of files.
- [x] *All* future TODOs are captured in issues, which are referenced in code comments.

I will need to create the TODO I mentioned above when I get back.

- [x] These code changes follow the ScubaGear [content style guide]
- [x] Related issues these changes resolve are linked preferably via [closing keywords]
- [x] All relevant type-of-change labels added.
- [x] All relevant project fields are set.
- [x] All relevant functional tests passed.

Actually, the Entra ID functional tests are all failing at this time.  But Ted is aware of the issues and is actively working them.

- [x] All automated checks (e.g., linting, static analysis, unit/smoke tests) passed.

## ✅ Pre-merge checklist ##

- [x] PR passed smoke test check.
- [x] Feature branch has been rebased against changes from parent branch, as needed
- [x] Resolved all merge conflicts on branch
- [x] Notified merge coordinator that PR is ready for merge via comment mention

## ✅ Post-merge checklist ##

- [ ] Feature branch deleted after merge to clean up repository.
- [ ] Verified that all checks pass on parent branch (e.g., main or release-name) after merge.

### CODE REVIEW

**Comment by james-garriss** [comment]
> As for cache existence, while cleanup is good, perhaps instead of always, only have it cleanup if all tests pass to facilitate re-running of failed jobs.

Done.

File: .github/workflows/test_production_function.yaml [hunk 1]
```diff
@@ -6,206 +6,166 @@
...
+      - ".github/workflows/function_test_defender.yaml"
+      - ".github/workflows/function_test_entra.yaml"
+      - ".github/workflows/function_test_exchange.yaml"
+      - ".github/workflows/function_test_powerplatform.yaml"
+      - ".github/workflows/function_test_sharepoint.yaml"
+      - ".github/workflows/function_test_teams.yaml"
+      - ".github/actions/extract-email/action.yaml"
+      - ".github/actions/get-thumbprint/action.yaml"
+      - ".github/actions/import-pfx/action.yaml"
+      - ".github/actions/test-product/action.yaml"
```
[the remark this answers is not in the document]
Comment (james-garriss) [reply .github/workflows/test_production_function.yaml:40]: Yes, those should be commented out just before merging to main.

File: .github/workflows/clear_cache.yaml [hunk 2]
```diff
@@ -0,0 +1,41 @@
+# Purpose: The purpose of this workflow is to delete all caches in the repo, as it has limited total cache storage.
+# Warning: This can disrupt other workflows that are using caches. Run with caution.
```
[the remark this answers is not in the document]
Comment (james-garriss) [reply .github/workflows/clear_cache.yaml:2]: That was the original comment, which was true when I created the original version of this workflow.  Updated to show its current constraints.

File: .github/actions/import-pfx/action.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+runs:
+  using: "composite"
+  steps:
+    - run: |
+        cd repo
+        # Create PFX
+        New-Item -ItemType file -Path ./key.txt
+        Set-Content -Path ./key.txt -Value "${{ inputs.PfxBase64 }}"
+        certutil -decode ./key.txt ./key.pfx
+        # Import FPX
```
[the remark this answers is not in the document]
Comment (james-garriss) [reply .github/actions/import-pfx/action.yaml:19]: Done.

File: .github/workflows/clear_cache.yaml
[the remark this answers is not in the document]
Comment (james-garriss) [reply .github/workflows/clear_cache.yaml:1]: I have taken a TODO to create a new issue to create this alternate approach.

File: .github/workflows/function_test_defender.yaml [hunk 4]
```diff
@@ -0,0 +1,572 @@
...
+      - name: Restore Dependencies
+        uses: actions/cache/restore@v4
+        with:
+          path: C:\Users\runneradmin\Documents\WindowsPowerShell\Modules
+          key: powershell-directory-
+      - name: Restore OPA
+        uses: actions/cache/restore@v4
+        with:
+          path: C:\Users\runneradmin\.scubagear\Tools
+          key: opa-directory-
```
[the remark this answers is not in the document]
Comment (james-garriss) [reply .github/workflows/function_test_defender.yaml:80]: If Publish fails, then the workflow itself fails and the test jobs do not run.

The caches are no longer always deleted, so there should not be any problem re-running jobs.

Thus, the only way this is an issue is if Publish succeeds and someone manages to delete those caches before for the test job runs.  This would have to be a very intentional action on someone's part, not something that could occur normally.  While this is possible, I think this is an extreme edge case that is not a legit item of concern.  I vote for doing nothing or perhaps putting an item in the backlog to handle this edge case at a later date (when we don't have more important things to do).

**Comment by james-garriss** [comment]
> I think the "inputs.GitHubEventName == 'push'" should be removed prior to the merge as it seems like it is for debugging.

This enables the email notification to happen when a functional test fails and the test was initiated by committing code.  Yes, that is part of debugging.  This important little piece of code was poorly documented and hard to find, so I don't want to delete them, so I commented them out and marked them as debugging.  This will be annoying whenever I have to modify and debug this code in the future, but this little "or" won't be running in production.

> Future TODO add all the tenants which fail to the list and only send 1 email to TCO listing all the failed tenants rather than spamming the TCOs if the multiple tenants fail. This could also help reduce repeated code for sending the email.

I'm not sure if that's possible, but it might be a good idea.  I added the idea to an existing TODO that I have in my backlog.

> Future TODO not sure if this is possible but an idea I had was to make a template to make the email body be a bit more specific or include links, or maybe be the same template as current GitHub Emails?

GitHub does not have a notification system, so this workflow is using an external notification that comes from Gmail.  For security reasons, the less it says the better.  I already have an existing TODO to consider if a link to the specific workflow can be added.

> You added a comment for "This allows someone to run tests manual and enter their own email" but I didn't see how to add your own email while running the branch, is that a future TODO?

This has already been implemented.  Look at the input on lines 14-23:
https://github.com/cisagov/ScubaGear/blob/1278-fix-concurrency-issues-for-nightly-functional-tests/.github/workflows/test_production_function.yaml

File: .github/workflows/clear_cache.yaml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+name: Clear Caches
+
+on:
+  workflow_call:
+  workflow_dispatch:
+
+# NOTE: This skip is not honored by Checkov.
+# checkov:skip=CKV2_GHA_1:Write permissions required to clear cache.
+# permissions: write-all
```
[the remark this answers is not in the document]
Comment (james-garriss) [reply .github/workflows/clear_cache.yaml:11]: Done.

**Comment by james-garriss** [comment]
Fixed bug with schedule that caused the email notification to occur even when run manually and the input not checked.  Should be good to go.

File: .github/workflows/test_production_function.yaml [hunk 6]
```diff
@@ -6,206 +6,172 @@
...
 name: Functional Tests
 
-# Run this workflow at 12:15 am
-# on every Sun to Thr (b/c GMT -5)
 on:
+  # Run this workflow at 12:15 am
+  # on every Sun to Thr (b/c GMT -5)
   schedule:
     - cron: "15 4 * * 0-4"
+  # This allows someone to run tests manual and enter their own email,
```
Comment (dagarwal-mitre) [inline .github/workflows/test_production_function.yaml:14]: This comment is a bit misleading, as currently I can't enter my email to get the failure emails, I can only turn on or off sending the emails to TCOs

**Review by dagarwal-mitre — CHANGES_REQUESTED** [review]
Overall looks good, just had a minor comment

**Review by dagarwal-mitre — APPROVED** [review]
Looks good!

**Comment by james-garriss** [comment]
@nanda-katikaneni , I think it's ready.
